# DS Automation Assignment

Using our prepared churn data from week 2:
- use pycaret to find an ML algorithm that performs best on the data
    - Choose a metric you think is best to use for finding the best model; by default, it is accuracy but it could be AUC, precision, recall, etc. The week 3 FTE has some information on these different metrics.
- save the model to disk
- create a Python script/file/module with a function that takes a pandas dataframe as an input and returns the probability of churn for each row in the dataframe
    - your Python file/function should print out the predictions for new data (new_churn_data.csv)
    - the true values for the new data are [1, 0, 0, 1, 0] if you're interested
- test your Python module and function with the new data, new_churn_data.csv
- write a short summary of the process and results at the end of this notebook
- upload this Jupyter Notebook and Python file to a Github repository, and turn in a link to the repository in the week 5 assignment dropbox

*Optional* challenges:
- return the probability of churn for each new prediction, and the percentile where that prediction is in the distribution of probability predictions from the training dataset (e.g. a high probability of churn like 0.78 might be at the 90th percentile)
- use other autoML packages, such as TPOT, H2O, MLBox, etc, and compare performance and features with pycaret
- create a class in your Python module to hold the functions that you created
- accept user input to specify a file using a tool such as Python's `input()` function, the `click` package for command-line arguments, or a GUI
- Use the unmodified churn data (new_unmodified_churn_data.csv) in your Python script. This will require adding the same preprocessing steps from week 2 since this data is like the original unmodified dataset from week 1.

In [2]:
from pycaret.classification import setup, compare_models, predict_model, save_model, load_model

In [5]:
import pandas as pd

In [6]:
df = pd.read_csv(r'C:\Users\Joseph\OneDrive\Desktop\Data Science\MSDS600\Week 5\Assignment\prepared_churn_data.csv', index_col='customerID')
df

tenure  PhoneService  Contract  PaymentMethod  MonthlyCharges  \
customerID                                                                  
7590-VHVEG     1.0             0         0              3           29.85   
5575-GNVDE    34.0             1         1              2           56.95   
3668-QPYBK     2.0             1         0              2           53.85   
7795-CFOCW    45.0             0         1              1           42.30   
9237-HQITU     2.0             1         0              3           70.70   
...            ...           ...       ...            ...             ...   
6840-RESVB    24.0             1         1              2           84.80   
2234-XADUH    72.0             1         1              0          103.20   
4801-JZAZL    11.0             0         0              3           29.60   
8361-LTMKD     4.0             1         0              2           74.40   
3186-AJIEK    66.0             1         2              1          105.65   

            TotalCharges  Churn  charge_per_tenure  
customerID                                          
7590-VHVEG         29.85      0          29.850000  
5575-GNVDE       1889.50      0          55.573529  
3668-QPYBK        108.15      1          54.075000  
7795-CFOCW       1840.75      0          40.905556  
9237-HQITU        151.65      1          75.825000  
...                  ...    ...                ...  
6840-RESVB       1990.50      0          82.937500  
2234-XADUH       7362.90      0         102.262500  
4801-JZAZL        346.45      0          31.495455  
8361-LTMKD        306.60      1          76.650000  
3186-AJIEK       6844.50      0         103.704545  

[7043 rows x 8 columns]

In [7]:
#Set up autoML
automl = setup(df, target='Churn')

In [9]:
automl.get_metrics()

Name Display Name  \
ID                                  
acc         Accuracy     Accuracy   
auc              AUC          AUC   
recall        Recall       Recall   
precision  Precision        Prec.   
f1                F1           F1   
kappa          Kappa        Kappa   
mcc              MCC          MCC   

                                              Score Function  \
ID                                                             
acc          <function accuracy_score at 0x00000253CC3A9A80>   
auc        <pycaret.internal.metrics.BinaryMulticlassScor...   
recall     <pycaret.internal.metrics.BinaryMulticlassScor...   
precision  <pycaret.internal.metrics.BinaryMulticlassScor...   
f1         <pycaret.internal.metrics.BinaryMulticlassScor...   
kappa      <function cohen_kappa_score at 0x00000253CC3A9...   
mcc        <function matthews_corrcoef at 0x00000253CC3A9...   

                                                      Scorer      Target  \
ID                                                                         
acc                                                 accuracy        pred   
auc        make_scorer(roc_auc_score, needs_proba=True, e...  pred_proba   
recall           make_scorer(recall_score, average=weighted)        pred   
precision     make_scorer(precision_score, average=weighted)        pred   
f1                   make_scorer(f1_score, average=weighted)        pred   
kappa                         make_scorer(cohen_kappa_score)        pred   
mcc                           make_scorer(matthews_corrcoef)        pred   

                                                    Args  Greater is Better  \
ID                                                                            
acc                                                   {}               True   
auc        {'average': 'weighted', 'multi_class': 'ovr'}               True   
recall                           {'average': 'weighted'}               True   
precision                        {'average': 'weighted'}               True   
f1                               {'average': 'weighted'}               True   
kappa                                                 {}               True   
mcc                                                   {}               True   

           Multiclass  Custom  
ID                             
acc              True   False  
auc              True   False  
recall           True   False  
precision        True   False  
f1               True   False  
kappa            True   False  
mcc              True   False

In [10]:
#use pycaret to find an ML algorithm that performs best on the data
best_model = compare_models()

In [11]:
best_model

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=1000,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=1470, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

Logistic Regression has the highest accuracy, so this algorithm will be used.

In [12]:
#save the model to disk
save_model(best_model, 'LR')

Transformation Pipeline and Model Successfully Saved


(Pipeline(memory=Memory(location=None),
          steps=[('numerical_imputer',
                  TransformerWrapper(exclude=None,
                                     include=['tenure', 'PhoneService',
                                              'Contract', 'PaymentMethod',
                                              'MonthlyCharges', 'TotalCharges',
                                              'charge_per_tenure'],
                                     transformer=SimpleImputer(add_indicator=False,
                                                               copy=True,
                                                               fill_value=None,
                                                               keep_empty_features=False,
                                                               missing_values=nan,
                                                               strategy='mean',
                                                               verbos...
           

In [21]:
#Create a Python script with a function that takes a pandas dataframe as an input and returns the probability of churn for each row in the dataframe
import pickle

with open('LR_model.pk', 'wb') as f:
    pickle.dump(best_model, f)

In [22]:
with open('LR_model.pk', 'rb') as f:
    loaded_model = pickle.load(f)

In [23]:
new_df = pd.read_csv(r'C:\Users\Joseph\OneDrive\Desktop\Data Science\MSDS600\Week 5\Assignment\new_churn_data.csv', index_col='customerID')
new_df

tenure  PhoneService  Contract  PaymentMethod  MonthlyCharges  \
customerID                                                                  
9305-CKSKC      22             1         0              2           97.40   
1452-KNGVK       8             0         1              1           77.30   
6723-OKKJM      28             1         0              0           28.25   
7832-POPKP      62             1         0              2          101.70   
6348-TACGU      10             0         0              1           51.15   

            TotalCharges  charge_per_tenure  
customerID                                   
9305-CKSKC        811.70          36.895455  
1452-KNGVK       1701.95         212.743750  
6723-OKKJM        250.90           8.960714  
7832-POPKP       3106.56          50.105806  
6348-TACGU       3440.97         344.097000

In [25]:
#Print out the predictions for new data
new_data = new_df
loaded_model.predict(new_data)

array([1, 0, 0, 0, 1], dtype=int8)

In [27]:
#Test Python module and function with the new data
predict_model(best_model,new_data)

tenure  PhoneService  Contract  PaymentMethod  MonthlyCharges  \
customerID                                                                  
9305-CKSKC      22             1         0              2       97.400002   
1452-KNGVK       8             0         1              1       77.300003   
6723-OKKJM      28             1         0              0       28.250000   
7832-POPKP      62             1         0              2      101.699997   
6348-TACGU      10             0         0              1       51.150002   

            TotalCharges  charge_per_tenure  prediction_label  \
customerID                                                      
9305-CKSKC    811.700012          36.895454                 1   
1452-KNGVK   1701.949951         212.743744                 0   
6723-OKKJM    250.899994           8.960714                 0   
7832-POPKP   3106.560059          50.105808                 0   
6348-TACGU   3440.969971         344.096985                 1   

            prediction_score  
customerID                    
9305-CKSKC            0.5011  
1452-KNGVK            0.5515  
6723-OKKJM            0.9294  
7832-POPKP            0.8447  
6348-TACGU            0.6486

# Summary

The first step involves installing PyCaret using conda package manager from the conda-forge channel. : The churn data is loaded from a CSV file (prepared_churn_data.csv) using pandas. Then, PyCaret's setup function is used to set up the classification experiment, specifying the target variable (Churn) and other parameterion: PyCaret's compare_models function is used to compare different machine learning modeCC. In this case, Logistic Regression (lr) is selected as the best m based on the accuracy scoreo ving: The best model (Logistic Regression) is saved to disk using PyCaret's save_model funictions: The saved model is loaded from disk using pickle. Then, new churn data is loaded from a CSV file (new_churn_data.csv). The loaded model is used to predict churn probabilities for the n Creation: Finally, a Python script/module is created with a function that takes a pandas DataFrame as input and returns the probability of churn for each row in the DataFrame. The function utilizes the saved model to make pr Overall, the model appears to be making predictions with varying levels of confidence ranging from 50%-93%.edictions.